cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,103 @@
+	// +kubebuilder:validation:Optional
+	// +optional
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec,casttype=CloudPrivateIPConfigSpec"`
+	// status is the observed status of the desired private IP request(s). Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status,casttype=CloudPrivateIPConfigStatus"`
+}
+
+// CloudPrivateIPConfigSpec specifies the requested list of private IPs and their corresponding node assignment.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpec struct {
+	// ipConfigs is a unique set of tuples (node, ip). A node can have several
+	// different IPs defined by creating a separate tuple for each.
+	// The order is not strict.
+	// +kubebuilder:validation:Required
+	// +required
+	IPConfigs []CloudPrivateIPConfigSpecItem `json:"ipConfigs" protobuf:"bytes,1,opt,name=ipConfigs"`
+}
+
+// CloudPrivateIPConfigStatus specifies the requested list of private IPs and their corresponding node assignment,
+// together with their assignment conditions.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatus struct {
+	// ipConfigs is the unique set of .spec.ipConfigs, where each is paired with an associated condition.
+	IPConfigs []CloudPrivateIPConfigCondition `json:"ipConfigs" protobuf:"bytes,1,opt,name=ipConfigs"`
+}
+
+// CloudPrivateIPConfigSpecItem is a tuple consisting of a node name and IP address.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigSpecItem struct {
+	// node is the node name, as specified by the Kubernetes field: node.metadata.name
+	// +kubebuilder:validation:Required
+	// +required
+	Node string `json:"node" protobuf:"bytes,1,opt,name=node"`
+	// ip is the specified IP address, can be IPv4 or IPv6
+	// +kubebuilder:validation:Required
+	// +required
+	IP string `json:"ip" protobuf:"bytes,2,opt,name=ip"`
+}
+
+// ConditionStatus specifies the current condition status of the CloudPrivateIPConfigCondition
+type ConditionStatus string
+
+const (
+	// Error defines a failed cloud assignment of the requested IP address.
+	Error ConditionStatus = "Error"
+
+	// Pending defines a pending cloud assignment of the requested IP address.
+	Pending ConditionStatus = "Pending"
+
+	// Success defines a successful cloud assignment of the requested IP address.
+	Success ConditionStatus = "Success"
+)
+
+// CloudPrivateIPConfigCondition is the assignment condition associated with each requested CloudPrivateIPConfigSpecItem
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigCondition struct {
+	CloudPrivateIPConfigSpecItem `json:",inline" protobuf:"bytes,1,opt,name=cloudPrivateIPConfigSpecItem"`
+	// status is the status of the condition.
+	// +kubebuilder:validation:Enum="Error";"Pending";"Success"
+	// +kubebuilder:validation:Required
+	// +required
+	Status ConditionStatus `json:"status" protobuf:"bytes,3,opt,name=status,casttype=ConditionStatus"`
```

Comment:
I would actually use a peer `Conditions []metav1.Condition` to aid in interoperability and allow flexibility of expression.  The fact that there is a `Pending` state, suggests that this is a multi-stage process.

Reply:
Yes, but I am running into a lot of loopholes when thinking about this. Essentially, we want to indicate an atomic status per item. If we apply a top level status as well, we'd have something along the lines of:

```
apiVersion: cloud.network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cluster
spec:
  ipConfigs:
  - node: nodeX
    ip: 192.168.126.11
  - node: nodeY
    ip: 192.168.126.102
  - node: nodeZ
    ip: 192.168.126.15
status:
  conditions:
  - message: "IP assignment: 192.168.126.102 failed"
    reason: AssignmentFailed
    status: "True"
    type: Failed
  - status: "True"
    type: Ready
  - message: "IP assignment: 192.168.126.15 is pending cloud API answer"
    reason: AssignmentPending
    status: "True"
    type: Pending
  ipConfigs:
  - node: nodeX
    ip: 192.168.126.11
    status: SUCCESS
    reason: ""
  - node: nodeY
    ip: 192.168.126.102
    status: ERROR
    reason: "Cloud API rejected the IP for some reason"
  - node: nodeZ
    ip: 192.168.126.15
    status: Pending
    reason: "Cloud API is processing the request"
```
That looks strange to me, I mean the top-level conditions state that it's both ready, not ready and pending....

What top level conditions would correspond and accurately reflect each atomic one? And we need an atomic status, because the network plugin needs to know each requested IP's status, so that it can proceed to programming the overlay accordingly.

Reply:
Given a potentially long running cloud API (implied by "Cloud API is processing the request"), I think that having conditions like

1. Progressing - True if some change is in progress and not yet done
2. Available - True if any egress up and functional
3. Degraded - True if any egress has been progressing more than 60 seconds and true if any egress change was rejected.

These are easy to pipe into a `oc wait` command and match the canonical status.